flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/optimize/program/FlinkRuntimeFilterProgram.java

```diff
@@ -0,0 +1,678 @@
+            // once runtime filter at the same place
+            return rel;
+        } else if (rel instanceof Exchange) {
+            // try to push the probe side to the input of exchange
+            Exchange exchange = (Exchange) rel;
+            return exchange.copy(
+                    exchange.getTraitSet(),
+                    Collections.singletonList(
+                            tryPushDownProbeAndInjectRuntimeFilter(
+                                    exchange.getInput(), probeIndices, buildSideInfo)));
+        } else if (rel instanceof Calc) {
+            // try to push the probe side to the input of projection
+            Calc calc = ((Calc) rel);
+            RexProgram program = calc.getProgram();
+            List<RexNode> projects =
+                    program.getProjectList().stream()
+                            .map(program::expandLocalRef)
+                            .collect(Collectors.toList());
+            ImmutableIntList inputIndices = getInputIndices(projects, probeIndices);
+            if (!inputIndices.isEmpty()) {
+                return calc.copy(
+                        calc.getTraitSet(),
+                        Collections.singletonList(
+                                tryPushDownProbeAndInjectRuntimeFilter(
+                                        calc.getInput(), inputIndices, buildSideInfo)));
+            }
+        } else if (rel instanceof Join) {
+            // try to push the probe side to the all inputs of join
+            Join join = (Join) rel;
+            Tuple2<ImmutableIntList, ImmutableIntList> tuple2 = getInputIndices(join, probeIndices);
+            ImmutableIntList leftIndices = tuple2.f0;
+            ImmutableIntList rightIndices = tuple2.f1;
+
+            if (!leftIndices.isEmpty() || !rightIndices.isEmpty()) {
+                RelNode leftSide = join.getLeft();
+                RelNode rightSide = join.getRight();
+
+                if (!leftIndices.isEmpty()) {
+                    leftSide =
+                            tryPushDownProbeAndInjectRuntimeFilter(
+                                    leftSide, leftIndices, buildSideInfo);
+                }
+
+                if (!rightIndices.isEmpty()) {
+                    rightSide =
+                            tryPushDownProbeAndInjectRuntimeFilter(
+                                    rightSide, rightIndices, buildSideInfo);
+                }
+
+                return join.copy(join.getTraitSet(), Arrays.asList(leftSide, rightSide));
+            }
+        } else if (rel instanceof BatchPhysicalGroupAggregateBase) {
+            // try to push the probe side to input of agg, iff the indices are all in grouping keys.
+            BatchPhysicalGroupAggregateBase agg = (BatchPhysicalGroupAggregateBase) rel;
+            int[] grouping = agg.grouping();
+            if (probeIndices.stream().allMatch(index -> (index < grouping.length))) {
```

Comment:
It seems that this judge's logic is not correct? The `index` is not necessarily smaller than `grouping.length`.

Reply:
The first `n(grouping.length)` elements of output of group agg are the group keys. In other words, if the index is greater than the `grouping.length`, it means it's not in the group key set, and cannot be pushed down further.

DPP  has the  [`similar logic`](https://github.com/apache/flink/blob/17be3716be42cd9dc5986e97ae222971d9d19be7/flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java#L414C1-L415C1)

Correct me if I'm wrong.

Reply:
Yes, you are right